File: lib/mercuryapi-1.23.0.20/cs/Samples/Codelets/ReadStopTrigger/ReadStopTrigger.cs
using System;
using System.Collections.Generic;
using System.Text;

// Reference the API
using ThingMagic;

namespace ReadStopTrigger
{
    /// <summary>
    /// Sample program that shows how to create a stoptrigger readplan to reads tags for a n number of tags
    /// and prints the tags found.
    /// </summary>
    class ReadStopTrigger
    {
        static void Main(string[] args)
        {
            // Program setup
            if (1 != args.Length)
            {
                Console.WriteLine(String.Join("\r\n", new string[] {
                    "Please provide reader URL, such as:",
                    "tmr:///com4",
                    "tmr://my-reader.example.com",
                }));
                Environment.Exit(1);
            }

            try
            {
                // Create Reader object, connecting to physical device.
                // Wrap reader in a "using" block to get automatic
                // reader shutdown (using IDisposable interface).
                using (Reader r = Reader.Create(args[0]))
                {
                    //Uncomment this line to add default transport listener.
                    //r.Transport += r.SimpleTransportListener;

                    r.Connect();
                    if (Reader.Region.UNSPEC == (Reader.Region)r.ParamGet("/reader/region/id"))
                    {
                        Reader.Region[] supportedRegions = (Reader.Region[])r.ParamGet("/reader/region/supportedRegions");
                        if (supportedRegions.Length < 1)
                        {
                            throw new FAULT_INVALID_REGION_Exception();
                        }
                        else
                        {
                            r.ParamSet("/reader/region/id", supportedRegions[0]);
                        }
                    }

                    // Set the q value
                    r.ParamSet("/reader/gen2/q", new Gen2.StaticQ(1));
                    
                    // Set the number of tags to read
                    StopOnTagCount sotc = new StopOnTagCount();
                    sotc.N = 1;

                    // Prepare multireadplan. Comment the below code if only single readplan is needed;
                    //StopTriggerReadPlan str1 = new StopTriggerReadPlan(sotc, new int[] { 1 }, 
                    //TagProtocol.GEN2, null, new Gen2.ReadData(Gen2.Bank.TID, 0, 0), 1000);
                    //StopTriggerReadPlan str2 = new StopTriggerReadPlan(sotc, new int[] { 2 }, 
                    //TagProtocol.GEN2, null, new Gen2.ReadData(Gen2.Bank.EPC, 0, 0), 1000);
                    //List<ReadPlan> plan = new List<ReadPlan>();
                    //plan.Add(str1);
                    //plan.Add(str2);
                    //MultiReadPlan readplan = new MultiReadPlan(plan);
                    
                    // Prepare single read plan. Comment the below code if multireadplan is needed
                    StopTriggerReadPlan readplan = new StopTriggerReadPlan(sotc, new int[] { 1 }, 
                        TagProtocol.GEN2, null, new Gen2.ReadData(Gen2.Bank.RESERVED, 0, 0), 1000);
                    
                    // Set readplan
                    r.ParamSet("/reader/read/plan", readplan);
                    
                    TagReadData [] tagReads;
                    // Read tags
                    tagReads = r.Read(1000);
                    // Print tag reads
                    foreach (TagReadData tr in tagReads)
                    {
                        Console.WriteLine(tr.ToString() + ", Protocol: " + tr.Tag.Protocol.ToString());
                        Console.WriteLine("Data: " + ByteFormat.ToHex(tr.Data));
                    }
                }
            }
            catch (ReaderException re)
            {
                Console.WriteLine("Error: " + re.Message);
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.Message);
            }
        }
    }
}

File: lib/mercuryapi-1.23.0.20/java/samples_nb/src/samples/ReadStopTrigger.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package samples;

import com.thingmagic.*;

/**
 *
 * Sample program that stops the read after finding N tags 
 * and prints the tags found.
 */
public class ReadStopTrigger {
    
     
  static void usage()
  {
    System.out.printf("Usage: demo reader-uri <command> [args]\n" +
                      "  (URI: 'tmr:///COM1' or 'tmr://astra-2100d3/' " +
                      "or 'tmr:///dev/ttyS0')\n\n" +
                      "Available commands:\n");
    System.exit(1);
  }

   public static void setTrace(Reader r, String args[])
  {    
    if (args[0].toLowerCase().equals("on"))
    {
      r.addTransportListener(r.simpleTransportListener);
    }    
  }
   

   static class TagReadListener implements ReadListener
  {
        public void tagRead(Reader r, TagReadData t) {
            System.out.println("Tag Read " + t);
        }
  }
  
  public static void main(String argv[])
  {
    // Program setup
    Reader r = null;
    int nextarg = 0;
    boolean trace = false;

    if (argv.length < 1)
      usage();

    if (argv[nextarg].equals("-v"))
    {
      trace = true;
      nextarg++;
    }
    
    // Create Reader object, connecting to physical device
    try
    { 
     
      TagReadData[] tagReads;
      
      r = Reader.create(argv[nextarg]);
      if (trace)
      {
        setTrace(r, new String[] {"on"});
      }
      r.connect();
      if (Reader.Region.UNSPEC == (Reader.Region)r.paramGet("/reader/region/id"))
      {
          Reader.Region[] supportedRegions = (Reader.Region[])r.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
          if (supportedRegions.length < 1)
          {
               throw new Exception("Reader doesn't support any regions");
      }
          else
          {
               r.paramSet("/reader/region/id", supportedRegions[0]);
          }
      }
      int initQ = 0;// set the Q value
      Gen2.StaticQ setQ = new Gen2.StaticQ(initQ);
      r.paramSet("/reader/gen2/q", setQ);
     
      StopOnTagCount sotc=new StopOnTagCount();
      sotc.N = 1; // number of tags to read
      StopTriggerReadPlan strp = new  StopTriggerReadPlan(sotc, new int[]{1},TagProtocol.GEN2);
      r.paramSet("/reader/read/plan", strp);
      r.addReadListener(new ReadStopTrigger.TagReadListener() );
      TagProtocol tagproto;
      // Read tags
      tagReads = r.read(1000);
      // Print tag reads
      for (TagReadData tr : tagReads)
      {
       System.out.println(tr.toString());
       tagproto= tr.getTag().getProtocol();
      }
      // Shut down reader
      r.destroy();
    } 
    catch (ReaderException re)
    {
      System.out.println("Reader Exception : " + re.getMessage());
    }
    catch (Exception re)
    {
      System.out.println("Exception : " + re.getMessage());
    }
  }
    
}
